text/0249-v2-experiments.md

```diff
@@ -0,0 +1,128 @@
+based on the CDK, but it does not address the backwards compatibility problem caused by minor releases with breaking
+changes to experimental modules. The current 2.0 plan does not include any consideration for the backwards compatibility
+issue.
+
+The CDK uses an exception to semantic versioning expectations by labeling certain modules as experimental, to allow us
+to make breaking changes to those modules in minor version upgrades. There is precedent for this in other open source
+projects, but for CDK users, it has been a constant source of pain and frustration. Users who do not carefully read and
+understand the documentation simply install packages, copy sample code, make a few tweaks and put the code into
+production. When they later upgrade to a new version, they are surprised to find that their code no longer works. The
+perception of instability has driven some users away, or caused them to limit their usage of CDK to the fundamental L1
+constructs, which do not provide them with the benefits of higher level abstractions.
+
+This RFC proposes that we stop experimenting in the main repository. To keep the implementation simple and to avoid
+possible inadvertent breakages, especially considering the need to convert from Typescript into a variety of languages,
+we should not use any clever techniques to simply hide the experiments. The code should be all stable. A user that
+installs aws-cdk-lib using npm or pip or any other package manager should be confident that there will be no intentional
+breaking changes in the 2.x line of releases for its lifetime.
+
+# README
+
+The following section will be removed from the CDK Readme:
+
+> Modules in the AWS Construct Library are designated Experimental while we build them; experimental modules may have
+> breaking API changes in any release. After a module is designated Stable, it adheres to semantic versioning, and only
+> major releases can have breaking changes. Each module's stability designation is available on its Overview page in the
+> AWS CDK API Reference. For more information, see Versioning in the CDK Developer Guide.
+
+The existing contributor guide does not mention experimental code, so the following will be added after the intro
+paragraph:
+
+> If you are contributing to a 2.x release, keep in mind that you may not introduce breaking changes to any public
+> facing API. Experimental modules are developed in a separate repository. [link]
+
+The new aws-cdk-previews repository will have the following Readme:
+
+> This repository contains experimental code related to the AWS Cloud Development Kit (AWS CDK). CDK is an open-source
+> software development framework to define cloud infrastructure in code and provision it through AWS CloudFormation.
+> Please refer to the main repository here [link]. [TODO...]
+
+The contributor guide can be copied from the main repository, since the toolchain for building the preview repository
+will be basically the same.
+
+# Design Summary
```

Comment:
Having read until this point and including the 'design summary' section, I still don't get how this is a better customer experience.

It seems that customers, now, instead of depending on just `aws-cdk-lib` will also take a dependency on `aws-cdk-previews` (or its individual packages). 

If `aws-cdk-previews` is going to follow (approximately) the same versioning scheme and release cycle, they're still going to see the same breaking changes.

Of course, they can choose not take a dependency on the `-previews` packages, but is that a practical option or how customers are going to behave? More likely than not, they will use the `-previews` packages, or resort to using the L1s. Now, we're back to the same problem again, right?

Did I miss something?

Reply:
Maybe the name should be more obvious, like `aws-cdk-experiments`, since one of our motivations is to solve for users who do not read the docs carefully.

Reply:
`aws-cdk-labs`?